### PR TITLE
Allow to dynamically change log filename using {dynamic name} segment

### PR DESCRIPTION
**Pull request by simon04**
### Description

Allow to dynamically change log filename. Say you have a task consisting of the steps foo, bar, baz, and you'd like to obtain log files `foo.log`, `bar.log`, `baz.log`

This PR would allow to:

```java
Configuration.set("writer.file", "logs/{dynamic name: init}.log");
Configuration.set("writer.policies", "dynamic name");
runInit(); // logs are written to logs/init.log
DynamicNameSegment.setDynamicName("foo");
runFoo(); // logs are written to logs/foo.log
DynamicNameSegment.setDynamicName("bar");
runBar(); // logs are written to logs/bar.log
DynamicNameSegment.setDynamicName("baz");
runBaz(); // logs are written to logs/baz.log
```

Would you be willing to support this feature? Is this draft implementation a feasible way to accomplish the goal? Is there a more concise way?

### Definition of Done

- [x] I read [contributing.md](https://github.com/tinylog-org/tinylog/blob/v2.4/contributing.md)
- [x] There are no TODOs left in the code
- [x] Code style follows the tinylog standard
- [x] All classes and methods have Javadoc
- [x] Changes are covered by JUnit tests including edge cases, errors, and exception handling
- [x] Maven build works including compiling, tests, and checks (`mvn verify`)
- [x] Changes are committed by a verified email address that is assigned to the GitHub account (https://github.com/settings/emails)

### Documentation

Additions or amendments for the [public documentation](https://github.com/tinylog-org/tinylog/wiki/Documentation):

```markdown

```

- [ ] document dynamic name segment including example code from above
- [ ] document dynamic name policy

### Agreements

- [x] I agree that my changes will be published under the terms of the [Apache License 2.0](https://github.com/tinylog-org/tinylog/blob/v2.0/license.txt) (mandatory)
- [x] I agree that my GitHub user name will be published in the release notes (optional)

### CODE REVIEW

**Comment by pmwmedia** [comment]
Thank you for your pull request, and yes, I'm willing o support this feature.

Improvement suggestions:

- Fix the code style
- Support an initial log name as parameter aka  `log file reset: my-log-file`
- `LogFileResetPolicy` does not need to be based on `DailyPolicy` (if you need both, you can configure it on runtime via `daily, log file reset`)
- I'm not completely happy with the names `LogFileResetPolicy` and `LogFileSegment`, but I don't know a better name by now

**Comment by simon04** [comment]
> * Fix the code style

Done.
 
> * Support an initial log name as parameter aka  `log file reset: my-log-file`

Done.

> * `LogFileResetPolicy` does not need to be based on `DailyPolicy` (if you need both, you can configure it on runtime via `daily, log file reset`)

Very nice!

> * I'm not completely happy with the names `LogFileResetPolicy` and `LogFileSegment`, but I don't know a better name by now

Me neither. I also dislike the cross dependency between the two classes. Is there a better way? Should one class implement both interfaces?

Related (and more elaborated) concepts in other logging libraries are:
- http://logback.qos.ch/manual/appenders.html#SiftingAppender
- https://logging.apache.org/log4j/2.x/manual/appenders.html#RoutingAppender

**Comment by pmwmedia** [comment]
You can fix the build by adding a bug pattern for `ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD` for `org.tinylog.policies.LogFileResetPolicy` into `configuration/spotbugs-filters.xml`.

**Comment by pmwmedia** [comment]
Some thoughts and suggestions:

- The actual log name could be stored in `LogFileSegment` instead of `LogFileResetPolicy`. This will make the `LogFileResetPolicy` a bit more complex, but allows to use the `LogFileSegment` without having a `LogFileResetPolicy`.
- Currently, my favorite naming would be `DynamicNameSegment` and `DynamicNamePolicy`. However, other ideas are welcome :)
- I have just noticed that JUnit tests for `LogFileResetPolicy` are missing.
- I thought about merging the segment and policy to a single class. However, I think having separate classes are better readable and maintainable.

**Comment by pmwmedia** [comment]
The release build fails because Java 6 does not support ` java.util.Objects.equals()`.

File: tinylog-impl/src/main/java/org/tinylog/path/DynamicNameSegment.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+package org.tinylog.path;
+
+import org.tinylog.policies.DynamicNamePolicy;
+import org.tinylog.policies.Policy;
+import org.tinylog.runtime.Timestamp;
+
+/**
+ * Path segment that represents the log filename from {@link DynamicNamePolicy}.
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/path/DynamicNameSegment.java]: I assume this description is outdated ;-)

File: tinylog-impl/src/main/java/org/tinylog/path/DynamicNameSegment.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	 * @return the current dynamic name
+	 */
+	public static String getDynamicName() {
+		return dynamicName;
+	}
+
+	/**
+	 * Sets a new dynamic name.
+	 * When used together with {@link DynamicNamePolicy} and the dynamic name differs from the current one,
+	 * a {@linkplain Policy#reset() reset} is triggered.
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/path/DynamicNameSegment.java]: Please place `<p>` tags around the second paragraph

File: tinylog-impl/src/main/java/org/tinylog/path/DynamicNameSegment.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	private static String dynamicName;
+
+	DynamicNameSegment(final String defaultValue) {
+		setDynamicName(defaultValue);
+	}
+
+	/**
+	 * Returns the current dynamic name.
+	 *
+	 * @return the current dynamic name
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/path/DynamicNameSegment.java]: Javadoc should always start with a upper case letter in tinylog

File: tinylog-impl/src/main/java/org/tinylog/path/DynamicNameSegment.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	public static String getDynamicName() {
+		return dynamicName;
+	}
+
+	/**
+	 * Sets a new dynamic name.
+	 * When used together with {@link DynamicNamePolicy} and the dynamic name differs from the current one,
+	 * a {@linkplain Policy#reset() reset} is triggered.
+	 *
+	 * @param newDynamicName the dynamic name to set
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/path/DynamicNameSegment.java]: Javadoc should always start with a upper case letter in tinylog

File: tinylog-impl/src/test/java/org/tinylog/policies/DynamicNamePolicyTest.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	 * Verifies that a warning will be output, if an argument is passed.
+	 */
+	@Test
+	public void warnIfArgumentIsSet() {
+		new DynamicNamePolicy("test");
+		assertThat(systemStream.consumeErrorOutput()).containsOnlyOnce("WARN").containsOnlyOnce("argument");
+	}
+
+	/**
+	 * Verifies that policy is registered as service under the name "startup".
```
Comment (pmwmedia) [inline tinylog-impl/src/test/java/org/tinylog/policies/DynamicNamePolicyTest.java]: Copy & paste mistake ;-)

**Review by pmwmedia — CHANGES_REQUESTED** [review]
The code looks perfect! I found only a view Javadoc issues.

**Review by pmwmedia — APPROVED** [review]

**Comment by pmwmedia** [comment]
Thank you for your changes. It looks perfect for me :)

**Comment by simon04** [comment]
Thank you for your patient and constructive reviews!

Sorry for a few glitches in Java 6 compatibility – I'm no longer used to write code targetting Java 6...
